File: 03-Stacks-and-Queues/07-Implementation-of-Loop_Queue/src/LoopQueue.java
/**
 * 避免出队, 所有元素移动, 使用循环队列避免元素移动
 * 循环队列, 当队尾无空间但队首有空间时, 将数据填充到队首
 *          但是, 不能填满, 要留一个位置
 * 空队, front == tail
 * 满队, (tail+1) % queue.length == front
 *
 * 复杂度
 *     void enqueue(E e);   O(1) 均摊
 *     E dequeue();         O(1) 均摊
 *     E getFront();        O(1)
 *     int getSize();       O(1)
 *     boolean isEmpty();   O(1)
 */
public class LoopQueue<E> implements Queue<E> {

    private E[] data;
    private int front,tail;
    private int size;

    public LoopQueue(int capacity) {
        data = (E[])(new Object[capacity+1]);
        front = 0;
        tail = 0;
        size = 0;
    }

    public LoopQueue(){
        this(10);
    }

    public int getCapacity() {
        return data.length-1;
    }

    @Override
    public boolean isEmpty() {
        return front == tail;
    }

    @Override
    public int getSize() {
        return size;
    }


    @Override
    public void enqueue(E e) {
        // 队列满了,扩容
        if ((tail+1) % data.length == front)
            resize(getCapacity()*2);

        data[tail] = e;
        tail = (tail+1) % data.length;

        size++;
    }

    @Override
    public E dequeue() {
        if (isEmpty())
            throw new IllegalArgumentException("Cannot dequeue from an empty queue.");

        E ret = data[front];
        data[front] = null;
        front = (front+1) % data.length;
        size--;

        // 缩容
        if (size == getCapacity() / 4 && getCapacity()/2 != 0)
            resize(getCapacity()/2);

        return ret;
    }

    @Override
    public E getFront() {
        if (isEmpty())
            throw new IllegalArgumentException("Cannot dequeue from an empty queue.");

        return data[front];
    }

    public void resize(int newCapacity) {
        E[] newData = (E[])new Object[newCapacity+1];
        for (int i = 0; i < size; i++)
            newData[i] = data[(i+front)%data.length];

        data = newData;
        front = 0;
        tail = size;
    }

    @Override
    public String toString() {
        StringBuilder res = new StringBuilder();
        res.append(String.format("Queue: size = %d, capacity = %d\n", size, getCapacity()));
        res.append("front [");
        for (int i=front; i!=tail; i=(i+1)%data.length) {
            res.append(data[i]);
            if ((i+1)%data.length != tail)
                res.append(", ");
        }
        res.append("] tail");

        return res.toString();
    }

    public static void main(String[] args) {

        LoopQueue<Integer> queue = new LoopQueue<>(4);
        for(int i = 0 ; i < 10 ; i ++){
            queue.enqueue(i);
            System.out.println(queue);
            if(i % 3 == 2){
                queue.dequeue();
                System.out.println(queue);
            }
        }
    }
}

File: go/03-Stacks-and-Queues/06-Loop-Queue/loopqueue/loopqueue.go
package loopqueue

import (
	"bytes"
	"fmt"
)

type LoopQueue struct {
	data  []interface{}
	size  int
	front int
	tail  int
}

func New(capacity int) *LoopQueue {
	return &LoopQueue{
		data: make([]interface{}, capacity+1),
	}
}

func (lq *LoopQueue) GetCapacity() int {
	return len(lq.data) - 1
}

func (lq *LoopQueue) GetSize() int {
	return lq.size
}
func (lq *LoopQueue) IsEmpty() bool {
	return lq.front == lq.tail
}
func (lq *LoopQueue) Enqueue(e interface{}) {
	// 队满
	if (lq.tail+1)%len(lq.data) == lq.front {
		lq.resize(lq.GetCapacity() * 2)
	}
	lq.data[lq.tail] = e
	lq.tail = (lq.tail + 1) % len(lq.data)
	lq.size++
}
func (lq *LoopQueue) Dequeue() interface{} {
	// 队空
	if lq.IsEmpty() {
		panic("cannot dequeue from empty queue")
	}
	e := lq.data[lq.front]
	lq.data[lq.front] = nil
	lq.front = (lq.front + 1) % len(lq.data)
	lq.size--

	if lq.size == lq.GetCapacity()/4 && lq.GetCapacity()/2 != 0 {
		lq.resize(lq.GetCapacity() / 2)
	}
	return e
}

func (lq *LoopQueue) GetFront() interface{} {
	if lq.IsEmpty() {
		panic("Queue is empty")
	}
	return lq.data[lq.front]
}

func (lq *LoopQueue) resize(capacity int) {
	newData := make([]interface{}, capacity)
	for i := 0; i < lq.size; i++ {
		newData[i] = lq.data[(i+lq.front)%len(lq.data)]
	}
	lq.data = newData
	lq.front = 0
	lq.tail = lq.size
}

func (lq *LoopQueue) String() string {
	var buffer bytes.Buffer

	buffer.WriteString(fmt.Sprintf("Queue: size = %d, capacity = %d\n", lq.size, lq.GetCapacity()))
	buffer.WriteString("front [")
	for i := lq.front; i != lq.tail; i = (i + 1) % len(lq.data) {
		// fmt.Sprint 将 interface{} 类型转换为字符串
		buffer.WriteString(fmt.Sprint(lq.data[i]))
		if (i+1)%len(lq.data) != lq.tail {
			buffer.WriteString(", ")
		}
	}
	buffer.WriteString("] tail")

	return buffer.String()
}
